samples/native_training/main.c

```diff
@@ -0,0 +1,215 @@
+    iree_runtime_session_device(session),
+    iree_hal_buffer_view_buffer(result), 0, &b,
+    sizeof(b), IREE_HAL_TRANSFER_BUFFER_FLAG_DEFAULT,
+    iree_infinite_timeout()));
+
+  // Update loss
+  IREE_RETURN_IF_ERROR(iree_runtime_call_outputs_pop_front_buffer_view(
+    &call, &result));
+  IREE_RETURN_IF_ERROR(iree_hal_device_transfer_d2h(
+    iree_runtime_session_device(session),
+    iree_hal_buffer_view_buffer(result), 0, &loss,
+    sizeof(loss), IREE_HAL_TRANSFER_BUFFER_FLAG_DEFAULT,
+    iree_infinite_timeout()));
+  iree_hal_buffer_view_release(result);
+
+  return iree_ok_status();
+}
+
+iree_status_t run_sample(iree_string_view_t bytecode_module_path,
+                         iree_string_view_t driver_name) {
+  iree_status_t status = iree_ok_status();
+
+  //===-------------------------------------------------------------------===//
+  // Instance configuration (this should be shared across sessions).
+  fprintf(stdout, "Configuring IREE runtime instance and '%s' device\n",
+          driver_name.data);
+  iree_runtime_instance_options_t instance_options;
+  iree_runtime_instance_options_initialize(&instance_options);
+  iree_runtime_instance_options_use_all_available_drivers(&instance_options);
+  iree_runtime_instance_t* instance = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_instance_create(&instance_options,
+                                          iree_allocator_system(), &instance);
+  }
+  // TODO(#5724): move device selection into the compiled modules.
+  iree_hal_device_t* device = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_instance_try_create_default_device(
+        instance, driver_name, &device);
+  }
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Session configuration (one per loaded module to hold module state).
+  fprintf(stdout, "Creating IREE runtime session\n");
+  iree_runtime_session_options_t session_options;
+  iree_runtime_session_options_initialize(&session_options);
+  iree_runtime_session_t* session = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_session_create_with_device(
+        instance, &session_options, device,
+        iree_runtime_instance_host_allocator(instance), &session);
+  }
+  iree_hal_device_release(device);
+
+  fprintf(stdout, "Loading bytecode module at '%s'\n",
+          bytecode_module_path.data);
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_session_append_bytecode_module_from_file(
+        session, bytecode_module_path.data);
+  }
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  fprintf(stdout, "Training...\n");
+
+  print_state();
+  for (int i = 0; i < 10; i++) {
+    IREE_RETURN_IF_ERROR(train(session));
+    print_state();
+  }
```

Comment:
So what would a user do after running this training? Write the learned weights to disk? Try inference on some sample inputs? It might help to log some additional summary information instead of just "Success!" after training completes, and/or include some comments about how to get data out of the program.

Reply:
I mentioned it in the README.  Does the explanation there make sense / need anything else?

Reply:
I would recommend that you add a check that the trained values are near the expected values or that the loss gets below some threshold so that the test actually checks the training procedure executes correctly.

Reply:
> I would recommend that you add a check that the trained values are near the expected values or that the loss gets below some threshold so that the test actually checks the training procedure executes correctly.

Done.